zeth_contracts/contracts/bls12_377_groth16.sol

```diff
@@ -0,0 +1,261 @@
+
+        assembly
+        {
+            // accum already in place
+
+            // Write g_2 to offset 0x080~
+            mstore(
+                add(pad, 0x080),
+                0x0000000000000000000000000000000000d6ac33b84947d9845f81a57a136bfa)
+            mstore(
+                add(pad, 0x0a0),
+                0x326e915fabc8cd6a57ff133b42d00f62e4e1af460228cd5184deae976fa62596)
+            mstore(
+                add(pad, 0x0c0),
+                0x0000000000000000000000000000000000b997fef930828fe1b9e6a1707b8aa5)
+            mstore(
+                add(pad, 0x0e0),
+                0x08a3dbfd7fe2246499c709226a0a6fef49f85b3a375363f4f8f6ea3fbd159f8a)
+            mstore(
+                add(pad, 0x100),
+                0x00000000000000000000000000000000002933c9ab1da3519734bb7d40c74f7c)
+            mstore(
+                add(pad, 0x120),
+                0x96f7cd46d372c68a05fbe4f5d29d09ebac0fdae50f6dde73818058280cc85ff1)
+            mstore(
+                add(pad, 0x140),
+                0x0000000000000000000000000000000000955cf57c9676d751f0b5431b46efdd)
+            mstore(
+                add(pad, 0x160),
+                0x5ea49e8f219e8b28d7bbd2176e29c49caa767e18b569d34dd5c5880f834819df)
+
+            // write vk.alpha and vk.minus_beta to offset 0x180~
+            mstore(add(pad, 0x180), sload(vk_slot_num))
+            mstore(add(pad, 0x1a0), sload(add(vk_slot_num,  1)))
+            mstore(add(pad, 0x1c0), sload(add(vk_slot_num,  2)))
+            mstore(add(pad, 0x1e0), sload(add(vk_slot_num,  3)))
+
+            mstore(add(pad, 0x200), sload(add(vk_slot_num,  4)))
+            mstore(add(pad, 0x220), sload(add(vk_slot_num,  5)))
+            mstore(add(pad, 0x240), sload(add(vk_slot_num,  6)))
+            mstore(add(pad, 0x260), sload(add(vk_slot_num,  7)))
+            mstore(add(pad, 0x280), sload(add(vk_slot_num,  8)))
+            mstore(add(pad, 0x2a0), sload(add(vk_slot_num,  9)))
+            mstore(add(pad, 0x2c0), sload(add(vk_slot_num, 10)))
+            mstore(add(pad, 0x2e0), sload(add(vk_slot_num, 11)))
+
+            // write proof.a and proof.b to offset 0x300~
+            mstore(add(pad, 0x300), mload(proof))
+            mstore(add(pad, 0x320), mload(add(proof, 0x020)))
+            mstore(add(pad, 0x340), mload(add(proof, 0x040)))
+            mstore(add(pad, 0x360), mload(add(proof, 0x060)))
+
+            mstore(add(pad, 0x380), mload(add(proof, 0x080)))
+            mstore(add(pad, 0x3a0), mload(add(proof, 0x0a0)))
+            mstore(add(pad, 0x3c0), mload(add(proof, 0x0c0)))
+            mstore(add(pad, 0x3e0), mload(add(proof, 0x0e0)))
+            mstore(add(pad, 0x400), mload(add(proof, 0x100)))
+            mstore(add(pad, 0x420), mload(add(proof, 0x120)))
+            mstore(add(pad, 0x440), mload(add(proof, 0x140)))
+            mstore(add(pad, 0x460), mload(add(proof, 0x160)))
+
+            // write proof.c, followed by vk.minus_delta to offset 0x480~
```

Comment:
What is the `~` at the end of the line? Do I understand correctly that this means "onward"? Same question for the lines above

Reply:
Yes.  I've used it in `0x300~0x380` to mean an interval, and in `0x300~` to mean onward.  I assumed it was pretty clear and intuitive, but if not we can change it.

Reply:
Thanks, was keen to double check to make sure I got it. Looks clear enough to me.